pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,840 @@
+	// need to remove all inline policies to be able to remove user
+	userPolicies, err := awsClient.ListUserPolicies(&iam.ListUserPoliciesInput{
+		UserName: aws.String(userName),
+	})
+	if err != nil {
+		return fmt.Errorf("error listing user inline policies: %v", err)
+	}
+	for _, policy := range userPolicies.PolicyNames {
+		_, err = awsClient.DeleteUserPolicy(&iam.DeleteUserPolicyInput{
+			UserName:   aws.String(userName),
+			PolicyName: policy,
+		})
+		if err != nil {
+			return fmt.Errorf("error deleting inline user policies: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// s3IamUserPolicy will attach iam policy to user granting r/w access to S3 bucketName
+func (c *Controller) s3IamUserPolicy(clusterID, userName, bucketName string, awsClient clustopaws.Client) error {
+	policyName := clusterID + "S3Access"
+	policyDocument, err := createS3IamPolicy(bucketName)
+	if err != nil {
+		return err
+	}
+
+	_, err = awsClient.PutUserPolicy(&iam.PutUserPolicyInput{
+		UserName:       aws.String(userName),
+		PolicyName:     aws.String(policyName),
+		PolicyDocument: aws.String(policyDocument),
+	})
+	if err != nil {
+		return fmt.Errorf("error attaching policy to user: %v", err)
+	}
+
+	return nil
+}
+
+func createS3IamPolicy(bucketName string) (string, error) {
+	t, err := template.New("s3policy").Parse(s3PolicyTemplate)
+	if err != nil {
+		return "", fmt.Errorf("error parsing s3policy template: %v", err)
+	}
+
+	params := s3PolicyTemplateParams{
+		BucketName: bucketName,
+	}
+
+	var buf bytes.Buffer
+	err = t.Execute(&buf, params)
+	if err != nil {
+		return "", fmt.Errorf("error processing s3policy template: %v", err)
+	}
+	policyDocument := buf.String()
+
+	return policyDocument, nil
+}
+
+func (c *Controller) syncRegistryInfra(cluster *clusteroperator.CombinedCluster, createInfra bool) error {
+	clusterID := cluster.Name
+	registrySecretAlreadyExists := false
+
+	// name for secret where we will store the registry object store credentials
+	objectStoreSecretName := controller.RegistryCredsSecretName(clusterID)
+
+	_, err := c.kubeClient.CoreV1().Secrets(cluster.Namespace).Get(objectStoreSecretName, metav1.GetOptions{})
+	secretFound := (err == nil)
+	unexpectedError := (err != nil && !errors.IsNotFound(err))
+
+	if unexpectedError {
+		return fmt.Errorf("error querying for existing registry secret: %v", err)
+	} else if secretFound && createInfra {
+		registrySecretAlreadyExists = true
+	}
+
+	bucketName := controller.RegistryObjectStoreName(clusterID)
+
+	awsClient, err := c.configureAws(cluster)
+	if err != nil {
+		return fmt.Errorf("error configuring AWS connection settings: %v", err)
+	}
+
+	if createInfra {
+		// create bucket if necessary
+		err = c.setupS3Bucket(bucketName, awsClient)
+		if err != nil {
+			return err
+		}
+
+		// create user if necessary
+		newAccessKey, err := c.setupS3IamUser(clusterID, bucketName, awsClient, registrySecretAlreadyExists)
```

Comment:
The fact that a secret with the creds already exists is not enough to decide whether to create new creds. We may have the secret in etcd, but someone may have deleted the user in AWS along with its creds. We should likely always create creds when creating a new user.